File: scripts/build_data.py
"""统一构建脚本：生成 arts_data.json 和 logo.js

功能：
1. 读取 scripts/arts_data.csv
2. 扫描 public/chararts/，给立绘条目标记 内置: true/false
3. 扫描 public/logos/，生成 logo.js
4. 生成 public/arts_data.json
"""

import csv
import json
import locale
import os
import sys
from datetime import datetime

# 初始化中文排序环境
if sys.platform == 'win32':
    try:
        locale.setlocale(locale.LC_COLLATE, 'chinese')
    except locale.Error:
        pass
else:
    try:
        locale.setlocale(locale.LC_COLLATE, 'zh_CN.UTF-8')
    except locale.Error:
        pass


def find_project_root():
    """从脚本所在目录向上查找项目根目录（包含 public/ 的目录）"""
    script_dir = os.path.dirname(os.path.abspath(__file__))
    candidate = script_dir
    for _ in range(3):
        if os.path.isdir(os.path.join(candidate, 'public')):
            return candidate
        candidate = os.path.dirname(candidate)
    # 回退：假设脚本在 scripts/ 下，项目根是其父目录
    return os.path.dirname(script_dir)


def scan_chararts(public_dir):
    """扫描 public/chararts/，返回内置立绘文件名集合"""
    chararts_dir = os.path.join(public_dir, 'chararts')
    if not os.path.isdir(chararts_dir):
        print(f"警告: chararts 目录不存在: {chararts_dir}")
        os.makedirs(chararts_dir, exist_ok=True)
        return set()
    files = {f for f in os.listdir(chararts_dir) if f.lower().endswith('.png')}
    print(f"扫描到 {len(files)} 个内置立绘文件")
    return files


def scan_logos(public_dir):
    """扫描 public/logos/，返回 logo 列表和扩展名映射"""
    logos_dir = os.path.join(public_dir, 'logos')
    if not os.path.isdir(logos_dir):
        print(f"警告: logos 目录不存在: {logos_dir}")
        os.makedirs(logos_dir, exist_ok=True)
        return [], {}

    logo_files = sorted(
        f for f in os.listdir(logos_dir)
        if f.lower().endswith('.png') or f.lower().endswith('.svg')
    )

    logo_list = [os.path.splitext(f)[0] for f in logo_files]
    # 按中文拼音排序
    logo_list.sort(key=locale.strxfrm)

    logo_ext_map = {}
    for f in logo_files:
        name, ext = os.path.splitext(f)
        logo_ext_map[name] = ext.lstrip('.').lower()

    print(f"扫描到 {len(logo_list)} 个Logo")
    return logo_list, logo_ext_map


def generate_logo_js(logo_list, logo_ext_map, src_data_dir):
    """生成 src/data/logo.js"""
    content = f"""// 自动生成的logo文件列表（logo名=文件名，不含后缀）

export const logos = {json.dumps(logo_list, ensure_ascii=False, indent=2)}

// logo名→文件扩展名映射（用于构建logo路径时判断.png/.svg）
export const logoExtMap = {json.dumps(logo_ext_map, ensure_ascii=False, indent=2)}
"""
    os.makedirs(src_data_dir, exist_ok=True)
    js_path = os.path.join(src_data_dir, 'logo.js')
    with open(js_path, 'w', encoding='utf-8') as f:
        f.write(content)
    print(f"已生成: {js_path} ({len(logo_list)} 个Logo)")


def convert_csv_to_json(csv_path, json_path, builtin_files):
    """读取 CSV 并生成 arts_data.json（含内置标记）"""
    with open(csv_path, 'r', encoding='utf-8-sig') as f:
        reader = csv.DictReader(f)
        required_fields = ['角色名', '外文名', '性别', '立绘编号', '文件名', '文件链接', 'logo', '出处']
        missing = [field for field in required_fields if field not in reader.fieldnames]
        if missing:
            sys.exit(f"CSV 缺少必填列: {missing}")
        rows = list(reader)

    chars = {}
    source_stats = {}
    char_source = {}

    for row in rows:
        name = row['角色名']
        source = row.get('出处', '')

        if name not in char_source:
            char_source[name] = source
            if source:
                if source not in source_stats:
                    source_stats[source] = {'角色数': 0, '立绘数': 0}
                source_stats[source]['角色数'] += 1

        if name not in chars:
            chars[name] = {
                '角色名': name,
                '外文名': row.get('外文名', ''),
                '性别': row.get('性别', ''),
                '立绘': [],
                'logo': row.get('logo', ''),
                '出处': source,
                '信息': {},
            }

        portrait_id = row.get('立绘编号', '')
        if any(p['编号'] == portrait_id for p in chars[name]['立绘']):
            print(f"警告: 角色 {name} 立绘编号 {portrait_id} 重复，已跳过")
        else:
            filename = row.get('文件名', '')
            is_builtin = filename in builtin_files
            chars[name]['立绘'].append({
                '编号': portrait_id,
                '文件名': filename,
                '文件链接': row.get('文件链接', ''),
                '内置': is_builtin,
            })
            if source and source in source_stats:
                source_stats[source]['立绘数'] += 1

        info = chars[name]['信息']
        if not info:
            info_fields = ['星级', '内部编号', '职业', '分支', '势力', '出身地']
            for field in info_fields:
                val = row.get(field, '').strip()
                if val:
                    if field == '星级':
                        try:
                            info[field] = int(val)
                        except ValueError:
                            info[field] = val
                    else:
                        info[field] = val

    # 统计内置立绘数
    builtin_count = sum(
        1 for c in chars.values() for p in c['立绘'] if p.get('内置')
    )
    print(f"内置立绘标记完成: {builtin_count} 张内置")

    # 检查孤儿文件：存在于 chararts/ 但未被任何 CSV 记录引用
    referenced_files = {p['文件名'] for c in chars.values() for p in c['立绘']}
    orphans = sorted(builtin_files - referenced_files)
    if orphans:
        print(f"\n[!] 发现 {len(orphans)} 个孤儿文件（存在于 chararts/ 但未被 CSV 引用）:")
        for f in orphans:
            print(f"  - {f}")
    else:
        print("无孤儿文件")

    # 构建角色数数组
    role_count_array = []
    total_portraits = sum(stats['立绘数'] for stats in source_stats.values())
    role_count_array.append({
        '出处': '不限',
        '角色数': len(chars),
        '立绘数': total_portraits
    })
    for source, stats in source_stats.items():
        role_count_array.append({
            '出处': source,
            '角色数': stats['角色数'],
            '立绘数': stats['立绘数']
        })

    output_data = {
        '元信息': {
            '转换时间': datetime.now().strftime('%Y-%m-%d %H:%M:%S'),
            '角色数': role_count_array
        },
        '角色': list(chars.values())
    }

    with open(json_path, 'w', encoding='utf-8') as f:
        json.dump(output_data, f, ensure_ascii=False, indent=2)

    print(f"已生成: {json_path} ({len(chars)} 个角色, {total_portraits} 张立绘)")
    print(f"统计: {role_count_array}")


def main():
    project_root = find_project_root()
    public_dir = os.path.join(project_root, 'public')
    src_data_dir = os.path.join(project_root, 'src', 'data')
    scripts_dir = os.path.join(project_root, 'scripts')

    csv_path = os.path.join(scripts_dir, 'arts_data.csv')
    json_path = os.path.join(public_dir, 'arts_data.json')

    if not os.path.isfile(csv_path):
        sys.exit(f"CSV 文件不存在: {csv_path}")

    print(f"项目根目录: {project_root}")
    print(f"读取 CSV: {csv_path}")
    print(f"输出 JSON: {json_path}")
    print()

    # 1. 扫描内置立绘
    builtin_files = scan_chararts(public_dir)

    # 2. 扫描 Logo，生成 logo.js
    logo_list, logo_ext_map = scan_logos(public_dir)
    generate_logo_js(logo_list, logo_ext_map, src_data_dir)

    # 3. 生成 arts_data.json（含内置标记）
    convert_csv_to_json(csv_path, json_path, builtin_files)

    print()
    print("构建完成！")


if __name__ == '__main__':
    main()

File: scripts/build_data.js
/**
 * 统一构建脚本：生成 arts_data.json 和 logo_data.json
 *
 * 功能：
 * 1. 读取 scripts/arts_data.csv 和 scripts/ark_chars.csv（两个来源的立绘数据）
 * 2. 扫描 public/chararts/，给立绘条目标记 内置: true/false
 * 3. 扫描 public/logos/，生成 logo_data.json
 * 4. 生成 public/arts_data.json
 */

import { readFileSync, writeFileSync, existsSync, mkdirSync, readdirSync } from 'fs';
import { join, dirname, extname } from 'path';
import { execSync } from 'child_process';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

/** 设为 false 可跳过方舟干员数据同步，加速调试 */
const SYNC_ARK_CHARS = false;

// --- CSV 解析 ---

function parseCSV(text) {
  // 移除 UTF-8 BOM
  if (text.charCodeAt(0) === 0xFEFF) {
    text = text.slice(1);
  }

  const rows = [];
  let row = [];
  let field = '';
  let inQuotes = false;

  for (let i = 0; i < text.length; i++) {
    const ch = text[i];

    if (inQuotes) {
      if (ch === '"') {
        if (text[i + 1] === '"') {
          field += '"';
          i++;
        } else {
          inQuotes = false;
        }
      } else {
        field += ch;
      }
    } else {
      if (ch === '"') {
        inQuotes = true;
      } else if (ch === ',') {
        row.push(field);
        field = '';
      } else if (ch === '\r') {
        row.push(field);
        field = '';
        rows.push(row);
        row = [];
        if (text[i + 1] === '\n') i++;
      } else if (ch === '\n') {
        row.push(field);
        field = '';
        rows.push(row);
        row = [];
      } else {
        field += ch;
      }
    }
  }

  // 最后一行
  if (field || row.length > 0) {
    row.push(field);
    rows.push(row);
  }

  return rows;
}

function csvDictReader(text) {
  const rows = parseCSV(text);
  if (rows.length === 0) return { fieldnames: [], records: [] };

  const fieldnames = rows[0];
  const records = [];

  for (let i = 1; i < rows.length; i++) {
    if (rows[i].length === 1 && rows[i][0] === '') continue; // 跳过空行
    const record = {};
    for (let j = 0; j < fieldnames.length; j++) {
      record[fieldnames[j]] = rows[i][j] !== undefined ? rows[i][j] : '';
    }
    records.push(record);
  }

  return { fieldnames, records };
}

// --- 项目路径 ---

function findProjectRoot() {
  /** 从脚本所在目录向上查找项目根目录（包含 public/ 的目录） */
  let candidate = __dirname;
  for (let i = 0; i < 3; i++) {
    if (existsSync(join(candidate, 'public'))) {
      return candidate;
    }
    candidate = dirname(candidate);
  }
  // 回退：假设脚本在 scripts/ 下，项目根是其父目录
  return dirname(__dirname);
}

// --- 扫描 ---

function scanChararts(publicDir) {
  /** 扫描 public/chararts/，返回内置立绘文件名集合 */
  const charartsDir = join(publicDir, 'chararts');
  if (!existsSync(charartsDir)) {
    console.log(`警告: chararts 目录不存在: ${charartsDir}`);
    mkdirSync(charartsDir, { recursive: true });
    return new Set();
  }
  const files = new Set(
    readdirSync(charartsDir).filter(f => f.toLowerCase().endsWith('.png'))
  );
  console.log(`扫描到 ${files.size} 个内置立绘文件`);
  return files;
}

function scanLogos(publicDir) {
  /** 扫描 public/logos/，返回 [{logo, ext}] 数组（按中文拼音排序） */
  const logosDir = join(publicDir, 'logos');
  if (!existsSync(logosDir)) {
    console.log(`警告: logos 目录不存在: ${logosDir}`);
    mkdirSync(logosDir, { recursive: true });
    return [];
  }

  const collator = new Intl.Collator('zh-CN');

  let logoFiles = readdirSync(logosDir).filter(
    f => f.toLowerCase().endsWith('.png') || f.toLowerCase().endsWith('.svg')
  );

  const logos = logoFiles.map(f => {
    const ext = extname(f);
    return { logo: f.slice(0, -ext.length), ext: ext.slice(1).toLowerCase() };
  });
  logos.sort((a, b) => collator.compare(a.logo, b.logo));

  console.log(`扫描到 ${logos.length} 个Logo`);
  return logos;
}

// --- 生成 ---

function generateLogoData(logos, publicDir) {
  /** 生成 public/logo_data.json */
  const jsonPath = join(publicDir, 'logo_data.json');
  writeFileSync(jsonPath, JSON.stringify(logos, null, 2), 'utf-8');
  console.log(`已生成: ${jsonPath} (${logos.length} 个Logo)`);
}

function convertCsvToJson(csvPaths, jsonPath, builtinFiles) {
  /** 读取多个 CSV 并生成 arts_data.json（含内置标记） */
  // 读取并合并所有 CSV 的记录
  let allRecords = [];
  let fieldnames = [];

  for (const csvPath of csvPaths) {
    if (!existsSync(csvPath)) {
      console.error(`CSV 文件不存在: ${csvPath}`);
      process.exit(1);
    }
    const text = readFileSync(csvPath, 'utf-8');
    const result = csvDictReader(text);
    if (fieldnames.length === 0) {
      fieldnames = result.fieldnames;
    }
    console.log(`读取 CSV: ${csvPath} (${result.records.length} 条记录)`);
    allRecords = allRecords.concat(result.records);
  }

  const requiredFields = ['角色名', '外文名', '性别', '立绘编号', '文件名', '文件链接', '势力logo', '出处'];
  const missing = requiredFields.filter(f => !fieldnames.includes(f));
  if (missing.length > 0) {
    console.error(`CSV 缺少必填列: ${JSON.stringify(missing)}`);
    process.exit(1);
  }

  const chars = {};
  const sourceStats = {};
  const charSource = {};

  for (const row of allRecords) {
    const name = row['角色名'];
    const source = row['出处'] || '';

    if (!(name in charSource)) {
      charSource[name] = source;
      if (source) {
        if (!(source in sourceStats)) {
          sourceStats[source] = { '角色数': 0, '立绘数': 0 };
        }
        sourceStats[source]['角色数'] += 1;
      }
    }

    if (!(name in chars)) {
      chars[name] = {
        '角色名': name,
        '外文名': row['外文名'] || '',
        '性别': row['性别'] || '',
        '立绘': [],
        'logo': row['势力logo'] || '',
        '出处': source,
        '信息': {},
      };
    }

    const portraitId = row['立绘编号'] || '';
    if (chars[name]['立绘'].some(p => p['编号'] === portraitId)) {
      console.log(`警告: 角色 ${name} 立绘编号 ${portraitId} 重复，已跳过`);
    } else {
      const filename = row['文件名'] || '';
      const isBuiltin = builtinFiles.has(filename);
      chars[name]['立绘'].push({
        '编号': portraitId,
        '文件名': filename,
        '文件链接': row['文件链接'] || '',
        '内置': isBuiltin,
        '时装品牌': row['时装品牌'] || '',
        '时装名': row['时装名'] || '',
      });
      if (source && source in sourceStats) {
        sourceStats[source]['立绘数'] += 1;
      }
    }

    const info = chars[name]['信息'];
    if (Object.keys(info).length === 0) {
      const infoFields = ['星级', '内部编号', '职业', '分支', '势力', '出身地', '上线时间'];
      for (const field of infoFields) {
        const val = (row[field] || '').trim();
        if (val) {
          if (field === '星级') {
            const num = parseInt(val, 10);
            info[field] = isNaN(num) ? val : num;
          } else {
            info[field] = val;
          }
        }
      }
    }
  }

  // 统计内置立绘数
  let builtinCount = 0;
  for (const c of Object.values(chars)) {
    for (const p of c['立绘']) {
      if (p['内置']) builtinCount++;
    }
  }
  console.log(`内置立绘标记完成: ${builtinCount} 张内置`);

  // 检查孤儿文件
  const referencedFiles = new Set();
  for (const c of Object.values(chars)) {
    for (const p of c['立绘']) {
      referencedFiles.add(p['文件名']);
    }
  }
  const orphans = [...builtinFiles].filter(f => !referencedFiles.has(f)).sort();
  if (orphans.length > 0) {
    console.log(`\n[!] 发现 ${orphans.length} 个孤儿文件（存在于 chararts/ 但未被 CSV 引用）:`);
    for (const f of orphans) {
      console.log(`  - ${f}`);
    }
  } else {
    console.log('无孤儿文件');
  }

  // 构建角色数数组
  const roleCountArray = [];
  const totalPortraits = Object.values(sourceStats).reduce((sum, s) => sum + s['立绘数'], 0);
  roleCountArray.push({
    '出处': '不限',
    '角色数': Object.keys(chars).length,
    '立绘数': totalPortraits,
  });
  for (const [source, stats] of Object.entries(sourceStats)) {
    roleCountArray.push({
      '出处': source,
      '角色数': stats['角色数'],
      '立绘数': stats['立绘数'],
    });
  }

  const outputData = {
    '元信息': {
      '转换时间': formatTimestamp(new Date()),
      '角色数': roleCountArray,
    },
    '角色': Object.values(chars),
  };

  writeFileSync(jsonPath, JSON.stringify(outputData, null, 2), 'utf-8');
  console.log(`已生成: ${jsonPath} (${Object.keys(chars).length} 个角色, ${totalPortraits} 张立绘)`);
  console.log(`统计: ${JSON.stringify(roleCountArray)}`);
}

function formatTimestamp(date) {
  const y = date.getFullYear();
  const M = String(date.getMonth() + 1).padStart(2, '0');
  const d = String(date.getDate()).padStart(2, '0');
  const h = String(date.getHours()).padStart(2, '0');
  const m = String(date.getMinutes()).padStart(2, '0');
  const s = String(date.getSeconds()).padStart(2, '0');
  return `${y}-${M}-${d} ${h}:${m}:${s}`;
}

// --- 主函数 ---

function main() {
  const projectRoot = findProjectRoot();
  const publicDir = join(projectRoot, 'public');
  const scriptsDir = join(projectRoot, 'scripts');

  const csvPaths = [
    join(scriptsDir, 'arts_data.csv'),
    join(scriptsDir, 'ark_chars.csv'),
  ];
  const jsonPath = join(publicDir, 'arts_data.json');

  for (const csvPath of csvPaths) {
    if (!existsSync(csvPath)) {
      console.error(`CSV 文件不存在: ${csvPath}`);
      process.exit(1);
    }
  }

  console.log(`项目根目录: ${projectRoot}`);
  console.log(`输出 JSON: ${jsonPath}`);
  console.log();

  // 0. 预先同步方舟干员数据
  if (SYNC_ARK_CHARS) {
    console.log('正在同步方舟干员数据...');
    try {
      execSync(`node ${join(scriptsDir, 'update_ark_chars.js')}`, { stdio: 'inherit', cwd: projectRoot });
    } catch (e) {
      console.error('同步方舟干员数据失败，继续使用已有数据:', e.message);
    }
    console.log();
  }

  // 1. 扫描内置立绘
  const builtinFiles = scanChararts(publicDir);

  // 2. 扫描 Logo，生成 logo_data.json
  const logos = scanLogos(publicDir);
  generateLogoData(logos, publicDir);

  // 3. 生成 arts_data.json（含内置标记）
  convertCsvToJson(csvPaths, jsonPath, builtinFiles);

  console.log();
  console.log('构建完成！');
}

main();
